### ISSUE TITLE
Issues if GPU > 2

### ISSUE THREAD

**Issue by Tom-Ryder** [BlackSamorez/tensor_parallel#98]
Hey - nice library!

I've noticed anything with GPU > 2 begins to run into issues. For example, a job that requires 30GB on one, might require 15GB on two. If you use  8, each GPU has 15GB. The benefit of the parallelism seems to stop beyond two. Any ideas? 

Thanks!

**Comment by BlackSamorez** [comment]
I've tested this library with 4 and 6 GPUs and everything looked fine. Could you, please, provide more information about what exactly you're doing and what your setup is?

**Comment by Tom-Ryder** [comment]
Here's a minimal example of the behaviour. I am working with 2d convolutions only, and the modifications to the config are designed to avoid shards with 0 elements (i.e. if number of devices > split channel dim). 

`import argparse
from typing import Sequence

import torch
import torch.nn as nn
from torch.nn.modules import conv

import tensor_parallel as tp
from tensor_parallel.config import Config
from tensor_parallel.state_actions import Scale, Split, SplitInsideChunks


def get_config(module: nn.Module, device_ids: Sequence[torch.device]) -> Config:
    state_rules = {}
    input_rules = {}
    output_rules = {}
    for name, module in module.named_modules():

        if isinstance(module, conv._ConvNd) and module.groups == 1:
            shape = [module.out_channels, module.in_channels] + list(module.kernel_size)
            shape[:2] = shape[:2][::-1] if module.transposed else shape[:2]
            shape = tuple(shape)
            
            if shape[0] < len(device_ids):
                split_dim = 1
            else:
                split_dim = 0

            assert module.weight.shape == shape, f"{module.weight.shape} != {shape}"
            assert module.bias is None or module.bias.shape == (module.out_channels,), module.bias.shape
            state_rules[f"^{name}.weight$"] = (
                Split(world_size=len(device_ids), dim=1)
                if module.transposed
                else Split(world_size=len(device_ids), dim=split_dim)
            )
            
            if module.bias is not None:
                if split_dim == 0:
                    state_rules[f"^{name}.bias$"] = Split(world_size=len(device_ids), dim=0)
                else:
                    state_rules[f"^{name}.bias$"] = Scale(world_size=len(device_ids))

            if split_dim == 0:
                output_rules[f"^{name}$"] = {0: "gather 1"}
            else:
                output_rules[f"^{name}$"] = {0: "sum"}
                
            if split_dim == 1:
                input_rules[f"^{name}$"] = {0: Split(world_size=len(device_ids), dim=split_dim)}

    return Config(state_rules, input_rules, output_rules, {})


class ResBlock(nn.Module):
    def __init__(self, dim):
        super().__init__()

        self.net = nn.Sequential(nn.Conv2d(dim, dim, 3, 1, 1),
                                 nn.ReLU(),
                                 nn.Conv2d(dim, dim, 3, 1, 1))

        self.act = nn.ReLU()

    def forward(self, x):
        skip = x
        out = self.act(self.net(x) + skip)    
        return out
    

class Model(nn.Module):
    def __init__(self, dim):
        super().__init__()
        self.net = nn.Sequential(
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),)
    
    def forward(self, x):
        return self.net(x)
    

if __name__ == "__main__":

    args_parser = argparse.ArgumentParser(description='List options for the experiment')
    args_parser.add_argument('--device_ids', nargs="+", default=[0, 1], type=int)
    
    args = args_parser.parse_args()

    model_base = Model(128)

    tp_config = get_config(model_base, args.device_ids)

    model = tp.TensorParallel(
        model_base,
        device_ids=args.device_ids,
        tensor_parallel_config=tp_config
        )

    x = torch.randn(1, 128, 512, 512).cuda(args.device_ids[0])

    for _ in range(1000):
        _ = model(x)`
        
       
With device_ids=[0, 1] I get ~2.4GB of RAM on two GPUs. With device_ids[0, 1, 2, 3, 4, 5, 6, 7] I still get ~2.4GB of RAM on each GPU across 8 (total memory 2.4 x 8). I hope I'm doing something stupid.

**Comment by Tom-Ryder** [comment]
Here's a minimal example of the behaviour. I am working with 2d convolutions only, and the modifications to the config are designed to avoid shards with 0 elements (i.e. if number of devices > split channel dim). 

```python import argparse
from typing import Sequence

import torch
import torch.nn as nn
from torch.nn.modules import conv

import tensor_parallel as tp
from tensor_parallel.config import Config
from tensor_parallel.state_actions import Scale, Split, SplitInsideChunks


def get_config(module: nn.Module, device_ids: Sequence[torch.device]) -> Config:
    state_rules = {}
    input_rules = {}
    output_rules = {}
    for name, module in module.named_modules():

        if isinstance(module, conv._ConvNd) and module.groups == 1:
            shape = [module.out_channels, module.in_channels] + list(module.kernel_size)
            shape[:2] = shape[:2][::-1] if module.transposed else shape[:2]
            shape = tuple(shape)
            
            if shape[0] < len(device_ids):
                split_dim = 1
            else:
                split_dim = 0

            assert module.weight.shape == shape, f"{module.weight.shape} != {shape}"
            assert module.bias is None or module.bias.shape == (module.out_channels,), module.bias.shape
            state_rules[f"^{name}.weight$"] = (
                Split(world_size=len(device_ids), dim=1)
                if module.transposed
                else Split(world_size=len(device_ids), dim=split_dim)
            )
            
            if module.bias is not None:
                if split_dim == 0:
                    state_rules[f"^{name}.bias$"] = Split(world_size=len(device_ids), dim=0)
                else:
                    state_rules[f"^{name}.bias$"] = Scale(world_size=len(device_ids))

            if split_dim == 0:
                output_rules[f"^{name}$"] = {0: "gather 1"}
            else:
                output_rules[f"^{name}$"] = {0: "sum"}
                
            if split_dim == 1:
                input_rules[f"^{name}$"] = {0: Split(world_size=len(device_ids), dim=split_dim)}

    return Config(state_rules, input_rules, output_rules, {})

class ResBlock(nn.Module):
    def __init__(self, dim):
        super().__init__()

        self.net = nn.Sequential(nn.Conv2d(dim, dim, 3, 1, 1),
                                 nn.ReLU(),
                                 nn.Conv2d(dim, dim, 3, 1, 1))

        self.act = nn.ReLU()

    def forward(self, x):
        skip = x
        out = self.act(self.net(x) + skip)    
        return out
    

class Model(nn.Module):
    def __init__(self, dim):
        super().__init__()
        self.net = nn.Sequential(
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),
                                ResBlock(dim),)
    
    def forward(self, x):
        return self.net(x)
    

if __name__ == "__main__":

    args_parser = argparse.ArgumentParser(description='List options for the experiment')
    args_parser.add_argument('--device_ids', nargs="+", default=[0, 1], type=int)
    
    args = args_parser.parse_args()

    model_base = Model(128)

    tp_config = get_config(model_base, args.device_ids)

    model = tp.TensorParallel(
        model_base,
        device_ids=args.device_ids,
        tensor_parallel_config=tp_config
        )

    x = torch.randn(1, 128, 512, 512).cuda(args.device_ids[0])

    for _ in range(1000):
        _ = model(x)
```
        
       
With device_ids=[0, 1] I get ~2.4GB of RAM on two GPUs. With device_ids[0, 1, 2, 3, 4, 5, 6, 7] I still get ~2.4GB of RAM on each GPU across 8 (total memory 2.4 x 8). I hope I'm doing something stupid.

**Comment by BlackSamorez** [comment]
I can reproduce this issue: memory footprint really is the same for 1, 2 and for GPUs.
I'll look into it.

**Comment by BlackSamorez** [comment]
It looks like the model itself is only 8mb. Despite that, at least 400mb of VRAM are allocated nevertheless.
I'd expect that either those numbers don't accurately represent real memory footprint or it's allocated in an inefficient manner because of a such small model size. It looks like it's not a `tensor_parallel` issue, because tensors are indeed being split correctly.
It definitely adds up for much larger models, so I guess there's nothing I can do about this particular case. It doesn't make sense to use `tensor_parallel` for such small models anyway.

**Comment by Tom-Ryder** [comment]
So my real use-case is a very large model, where a batch size of 1 will not fit on one A100, if that helps. But thanks for the efforts.

**Comment by BlackSamorez** [comment]
It definitely works as expected for larger models because all the demos would physically work otherwise.
I'm considering this issue resolved. If anything else comes up, feel free to reopen it!
